AP_LandingGear: add OPTIONS parameter and integrate with Copter

rmackay9 opened the pull request:
This PR adds an OPTIONS bitmask parameter that allows enabling automatic retracting and deploying of the gear in Auto, Guided and RTL modes.

This PR also adds support for:

- automatically retracting the landing gear after taking off in Guided
- deploying the landing gear during the final states of SmartRTL

This has been somewhat extensively tested in SITL:

- Ensured no change in behaviour before and after this PR in terms of the gear automatically retracting and deploying in Auto mode
- Checked LGR_OPTIONS = 1 caused the gear to be retracted but not deployed
- Checked LGR_OPTIONS = 2 caused the gear to be deployed but not retracted
- Checked the gear retracted and deployed appropriate in Guided and SmartRTL

Note: I have not integrated this new OPTIONS parameter into Plane.  I had a look but it wasn't completely straightforward.

Note2: this comes from [this discussion in the Copter-4.0 category](https://discuss.ardupilot.org/t/landing-gear-retract-on-auto-mission-start/51978) and also [comments from this PR](https://github.com/ArduPilot/ardupilot/pull/11033).

peterbarker commented on libraries/AP_LandingGear/AP_LandingGear.cpp:
```diff
@@ -67,6 +67,14 @@ const AP_Param::GroupInfo AP_LandingGear::var_info[] = {
+    // @Param: OPTIONS
+    // @DisplayName: Landing gear auto retract/deploy options
+    // @Description: Options to retract or deploy landing gear in Auto or Guided mode
+    // @Values: 1:Retract after Takeoff, 2:Deploy during Land, 3:Retract after Takeoff AND deploy during Land
```
Make this frame-specific if we're not going to do Plane.

rmackay9 replied:
I've raised an issue to add support for Plane as well. https://github.com/ArduPilot/ardupilot/issues/13665.  I think we will do this in the near future so I'll skip temporariliy making the param invisible for non-Copter frames

rmackay9 commented:
I've made two changes as discussed on the dev call:

- changed the two new methods so they cause the gear to retract or deploy based on the OPTIONS parameter.  This reduces the two step process we had previously in the PR (i.e. step1 was to check the accessor, step2 was to change the position of the landing gear)
- changed the auto deployment to be edge based rather than level based.

I've re-tested this in SITL and it works.

I did some extra testing and confirmed that in master, the pilot is unable to retract the landing gear during landing while after this PR they can.

Any further comments are greatly appreciated but if there aren't any then I'll probably go ahead and merge.  I've created an issue here to integrate with Plane as well: https://github.com/ArduPilot/ardupilot/issues/13665

rmackay9 commented:
This is included in Copter-4.0.4-rc1
